File: wrappers/python3/scripts/safe_archive.py
#!/usr/bin/env python3
"""Non-destructive archival tool for failure logs with M0-P1-I3 no-clobber semantics.

This module moves failure logs from .agent/FAIL-LOGS to .agent/FAIL-ARCHIVE
with configurable compression and strict no-clobber guarantees.

Purpose
-------
Provides safe, non-destructive archival of agent failure logs to prevent
accidental data loss while managing disk space through optional compression.
The tool never overwrites existing archive files unless explicitly configured
to fail (strict no-clobber mode).

M0-P1-I3 No-Clobber Semantics
------------------------------
Two modes of operation:

1. **Auto-suffix mode (default)**:
   - If destination exists, append .2, .3, etc. until unique name found
   - Never overwrites existing archive files
   - Emits WARNING to stderr when auto-suffixing occurs

2. **Strict no-clobber mode** (via --no-clobber flag or SAFE_ARCHIVE_NO_CLOBBER=1):
   - If destination exists, fail with exit code 2
   - Prevents any archival when collision detected
   - Use for workflows requiring explicit collision handling

Environment Variables
---------------------
SAFE_FAIL_DIR : str, optional
    Source directory for failure logs (default: .agent/FAIL-LOGS)

SAFE_ARCHIVE_DIR : str, optional
    Destination directory for archived logs (default: .agent/FAIL-ARCHIVE)

SAFE_ARCHIVE_COMPRESS : str, optional
    Compression method: none (default) | gzip | xz | zstd
    - none: No compression applied
    - gzip: Python built-in gzip compression (always available)
    - xz: Requires 'xz' command in PATH, uses multi-threading (-T0)
    - zstd: Requires 'zstd' command in PATH, uses multi-threading (-T0)

SAFE_ARCHIVE_NO_CLOBBER : str, optional
    Set to "1" to enable strict no-clobber mode globally

CLI Interface
-------------
    python3 safe_archive.py [--no-clobber] [--all | <file> ...]

Options:
    --no-clobber    Enable strict no-clobber mode (fail on collision)
    --all           Archive all files in SAFE_FAIL_DIR
    <file> ...      Archive specific file(s) by path

Examples
--------
Archive specific failure log::

    python3 safe_archive.py .agent/FAIL-LOGS/fail-2024-01-15T10-30-00.txt

Archive all failure logs with gzip compression::

    export SAFE_ARCHIVE_COMPRESS=gzip
    python3 safe_archive.py --all

Strict no-clobber mode via flag::

    python3 safe_archive.py --no-clobber mylog.txt
    # Fails with exit 2 if .agent/FAIL-ARCHIVE/mylog.txt exists

Auto-suffix mode (default)::

    python3 safe_archive.py mylog.txt
    # If mylog.txt exists, creates mylog.txt.2
    # If mylog.txt.2 exists, creates mylog.txt.3, etc.

Exit Codes
----------
0
    All files archived successfully
2
    Error: File not found, compression tool missing, or no-clobber collision

Side Effects
------------
- Creates SAFE_FAIL_DIR if it doesn't exist
- Creates SAFE_ARCHIVE_DIR if it doesn't exist
- Moves (not copies) source files to archive directory
- May compress archived files (removes uncompressed original)
- Prints status messages to stderr (ARCHIVED, WARNING, ERROR)

Filesystem Operations
---------------------
- Uses shutil.move() for atomic file operations
- Compressed files get extension: .gz (gzip), .xz (xz), .zst (zstd)
- Source file deleted after successful compression
- No partial writes (compression creates new file, then removes source)

Platform Notes
--------------
- gzip compression: Python built-in, available on all platforms
- xz compression: Requires 'xz' binary in PATH (common on Linux/macOS)
- zstd compression: Requires 'zstd' binary in PATH (install via package manager)
- Uses os.access(os.X_OK) to verify command availability

Contract References
-------------------
- **M0-P1-I3**: No-clobber archival semantics (auto-suffix and strict modes)

See Also
--------
- scripts/python3/safe_run.py: Generates failure logs for archival
- scripts/python3/safe_check.py: Verifies archival contract conformance
"""
import os
import sys
import shutil
import subprocess
import gzip
from typing import List


def eprint(*args: object) -> None:
    """Print to stderr for status and error messages.

    :param args: Variable arguments to print (passed to print())
    :returns: None

    All output is sent to stderr to avoid interfering with stdout,
    which may be captured or redirected by calling scripts.
    """
    print(*args, file=sys.stderr)


def usage() -> int:
    """Print usage message and return exit code 2.

    :returns: 2 (usage error exit code)

    Displays comprehensive help including options, environment variables,
    and examples. Caller is expected to exit with the returned code.
    """
    eprint("Usage: scripts/python3/safe_archive.py [--no-clobber] [--all | <file> ...]")
    eprint("")
    eprint("Options:")
    eprint("  --no-clobber            Fail if destination exists (default: auto-suffix)")
    eprint("")
    eprint("Environment:")
    eprint("  SAFE_FAIL_DIR           Source directory (default: .agent/FAIL-LOGS)")
    eprint("  SAFE_ARCHIVE_DIR        Destination directory (default: .agent/FAIL-ARCHIVE)")
    eprint("  SAFE_ARCHIVE_COMPRESS   Compression: none|gzip|xz|zstd (default: none)")
    eprint("  SAFE_ARCHIVE_NO_CLOBBER Set to 1 to enable strict no-clobber mode")
    return 2


def have_cmd(cmd: str) -> bool:
    """Check if a command is available in PATH.

    :param cmd: Command name to search for (e.g., "xz", "zstd")
    :returns: True if command exists and is executable, False otherwise

    Used to verify compression tool availability before attempting to
    invoke external commands for xz or zstd compression.

    Platform Notes
    --------------
    - Searches all directories in PATH environment variable
    - Requires execute permission (os.X_OK) in addition to existence
    - Does not verify command functionality, only availability
    """
    for p in os.environ.get("PATH", "").split(os.pathsep):
        exe = os.path.join(p, cmd)
        if os.path.isfile(exe) and os.access(exe, os.X_OK):
            return True
    return False


def compress_file(method: str, path: str) -> None:
    """Compress a file in-place using the specified method.

    :param method: Compression method: "none" | "gzip" | "xz" | "zstd"
    :param path: Path to file to compress (will be replaced with compressed version)
    :returns: None
    :raises RuntimeError: If method is invalid or compression tool not found

    Compression Methods
    -------------------
    none
        No compression applied, file unchanged
    gzip
        Python built-in gzip compression (always available)
        Creates path.gz, removes original
    xz
        External 'xz' command with multi-threading (-T0)
        Creates path.xz, removes original
        Requires 'xz' in PATH
    zstd
        External 'zstd' command with multi-threading (-T0)
        Creates path.zst, removes original
        Requires 'zstd' in PATH

    Side Effects
    ------------
    - Original file is deleted after successful compression
    - Compressed file has appropriate extension (.gz, .xz, .zst)
    - External commands invoked via subprocess.check_call()

    Error Handling
    --------------
    - Raises RuntimeError if compression tool not found in PATH
    - Raises RuntimeError if method is not one of the supported values
    - subprocess.check_call() will raise CalledProcessError on command failure

    Examples
    --------
    >>> compress_file("none", "mylog.txt")  # No-op
    >>> compress_file("gzip", "mylog.txt")  # Creates mylog.txt.gz, removes mylog.txt
    """
    method = method or "none"
    if method == "none":
        return
    if method == "gzip":
        out = path + ".gz"
        with open(path, "rb") as fin, gzip.open(out, "wb") as fout:
            shutil.copyfileobj(fin, fout)
        os.unlink(path)
        return
    if method == "xz":
        if not have_cmd("xz"):
            raise RuntimeError("xz command not found in PATH")
        subprocess.check_call(["xz", "-T0", "-f", path])
        return
    if method == "zstd":
        if not have_cmd("zstd"):
            raise RuntimeError("zstd command not found in PATH")
        subprocess.check_call(["zstd", "-q", "-T0", "-f", path])
        return
    raise RuntimeError(f"Invalid SAFE_ARCHIVE_COMPRESS value: {method}")


def archive_one(src: str, archive_dir: str, compress: str, strict_no_clobber: bool = False) -> None:
    """Archive a single file with M0-P1-I3 no-clobber semantics.

    :param src: Source file path
    :param archive_dir: Destination directory
    :param compress: Compression method
    :param strict_no_clobber: If True, fail if destination exists. If False (default), auto-suffix.
    """
    if not os.path.exists(src):
        raise RuntimeError(f"File not found: {src}")

    base = os.path.basename(src)
    dest = os.path.join(archive_dir, base)

    if os.path.exists(dest):
        if strict_no_clobber:
            # M0-P1-I3: Strict no-clobber mode - fail with error
            raise RuntimeError(f"Destination exists: {dest}")
        # M0-P1-I3: Default auto-suffix mode - append .2, .3, etc.
        n = 2
        while os.path.exists(dest + "." + str(n)):
            n += 1
        dest = dest + "." + str(n)
        eprint(f"WARNING: destination exists, using auto-suffix: {dest}")

    shutil.move(src, dest)
    eprint(f"ARCHIVED: {src} -> {dest}")
    compress_file(compress, dest)


def main(argv: List[str]) -> int:
    """Execute archival operation with M0-P1-I3 no-clobber semantics.

    :param argv: Command-line arguments (--no-clobber, --all, or file paths)
    :returns: Exit code (0 on success, 2 on error)
    :raises SystemExit: Via usage() on help request or missing arguments

    Argument Parsing
    ----------------
    - Detects -h/--help and calls usage()
    - Extracts --no-clobber flag if present
    - Checks SAFE_ARCHIVE_NO_CLOBBER environment variable
    - Requires either --all or one or more file paths

    Operation Modes
    ---------------
    --all
        Archives all files in SAFE_FAIL_DIR (sorted by name)
        Skips directories and symlinks

    <file> ...
        Archives specific files by path
        Files can be from any directory (not limited to SAFE_FAIL_DIR)

    Environment Configuration
    -------------------------
    - SAFE_FAIL_DIR: Source directory (default: .agent/FAIL-LOGS)
    - SAFE_ARCHIVE_DIR: Destination directory (default: .agent/FAIL-ARCHIVE)
    - SAFE_ARCHIVE_COMPRESS: Compression method (default: none)
    - SAFE_ARCHIVE_NO_CLOBBER: Set to "1" for strict no-clobber mode

    Side Effects
    ------------
    - Creates SAFE_FAIL_DIR if it doesn't exist
    - Creates SAFE_ARCHIVE_DIR if it doesn't exist
    - Moves source files to archive directory
    - Compresses archived files if SAFE_ARCHIVE_COMPRESS is set
    - Prints status messages to stderr (ARCHIVED, WARNING, ERROR)

    Error Handling
    --------------
    - Catches RuntimeError from archive_one() and compress_file()
    - Prints ERROR message to stderr
    - Returns exit code 2 on any error
    - --all mode: Continues processing remaining files after error (non-fatal)

    Examples
    --------
    >>> main(["--all"])  # Archive all files in SAFE_FAIL_DIR
    0

    >>> main(["--no-clobber", "mylog.txt"])  # Strict mode
    2  # If destination exists

    >>> main(["file1.txt", "file2.txt"])  # Archive specific files
    0
    """
    if not argv or argv[0] in ("-h", "--help"):
        return usage() if not argv else usage()

    # M0-P1-I3: Check for strict no-clobber mode
    strict_no_clobber = False
    args = list(argv)

    # Check for --no-clobber flag
    if "--no-clobber" in args:
        strict_no_clobber = True
        args.remove("--no-clobber")

    # Check for SAFE_ARCHIVE_NO_CLOBBER env var
    if os.environ.get("SAFE_ARCHIVE_NO_CLOBBER") == "1":
        strict_no_clobber = True

    fail_dir = os.environ.get("SAFE_FAIL_DIR", ".agent/FAIL-LOGS")
    archive_dir = os.environ.get("SAFE_ARCHIVE_DIR", ".agent/FAIL-ARCHIVE")
    compress = os.environ.get("SAFE_ARCHIVE_COMPRESS", "none")

    os.makedirs(fail_dir, exist_ok=True)
    os.makedirs(archive_dir, exist_ok=True)

    try:
        if args and args[0] == "--all":
            files = [
                os.path.join(fail_dir, n)
                for n in sorted(os.listdir(fail_dir))
                if os.path.isfile(os.path.join(fail_dir, n))
            ]
            if not files:
                eprint(f"No files to archive in {fail_dir}")
                return 0
            for f in files:
                archive_one(f, archive_dir, compress, strict_no_clobber)
            return 0

        if not args:
            return usage()

        for f in args:
            archive_one(f, archive_dir, compress, strict_no_clobber)
        return 0
    except RuntimeError as e:
        eprint(f"ERROR: {e}")
        return 2


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: wrappers/perl/scripts/safe_archive.pl
#!/usr/bin/env perl

=head1 NAME

safe-archive.pl - Non-destructive archival of failure logs with optional compression

=head1 SYNOPSIS

  # Archive all failure logs
  safe-archive.pl --all
  
  # Archive specific files
  safe-archive.pl <file1> [file2 ...]
  
  # Show help
  safe-archive.pl --help

=head1 DESCRIPTION

This script archives failure logs from the FAIL-LOGS directory to the
FAIL-ARCHIVE directory using no-clobber semantics. It never deletes or
overwrites existing files, ensuring safe archival operations.

The script supports optional compression (gzip, xz, zstd) and handles
filename collisions by automatically adding numeric suffixes.

=head1 OPTIONS

=over 4

=item B<--all>

Archive all files in the FAIL-LOGS directory. This is the most common usage.

=item B<-h>, B<--help>

Display usage information and exit with code 2.

=item B<E<lt>fileE<gt> ...>

Archive specific files by path. Files must exist or an error is raised.

=back

=head1 ARGUMENTS

=over 4

=item B<file1>, B<file2>, ...

One or more file paths to archive. Each file is moved from its current
location to the archive directory. Paths can be absolute or relative.

=back

=head1 ENVIRONMENT VARIABLES

=over 4

=item B<SAFE_FAIL_DIR>

Source directory containing failure logs. Default: C<.agent/FAIL-LOGS>

The script expects log files to exist in this directory.

=item B<SAFE_ARCHIVE_DIR>

Destination directory for archived logs. Default: C<.agent/FAIL-ARCHIVE>

The script creates this directory if it doesn't exist.

=item B<SAFE_ARCHIVE_COMPRESS>

Compression method to apply after archival. Valid values:

  - none (default) - No compression
  - gzip - Compress with gzip (requires IO::Compress::Gzip)
  - xz - Compress with xz (requires xz command in PATH)
  - zstd - Compress with zstd (requires zstd command in PATH)

Compression is applied AFTER the file is moved to the archive directory.
The original uncompressed file is removed after successful compression.

=back

=head1 EXIT CODES

=over 4

=item B<0>

Success - All files archived successfully.

=item B<1>

File operation error - Failed to move file, compress file, or file not found.
Error message printed to stderr.

=item B<2>

Usage error - Invalid arguments or help requested.

=back

=head1 NO-CLOBBER SEMANTICS

The archival process implements strict no-clobber behavior per M0-P1-I3:

=over 4

=item 1. B<Check destination>

If a file with the same basename exists in the archive directory, do NOT
overwrite it.

=item 2. B<Auto-suffix collision resolution>

Add a numeric suffix (1, 2, 3, ...) to create a unique filename:

  Original: my-log.log
  Collision detected: my-log.log exists
  New name: my-log.1.log
  
  If my-log.1.log also exists:
  New name: my-log.2.log

=item 3. B<Preserve original>

The existing file in the archive directory is never modified or deleted.

=item 4. B<Atomic move>

Files are moved (not copied) to ensure atomic filesystem operations.

=back

=head1 COMPRESSION SUPPORT

=head2 gzip

Requires: C<IO::Compress::Gzip> Perl module

Result: C<filename.log.gz>

The script attempts to load IO::Compress::Gzip at runtime. If not available,
an error is raised.

=head2 xz

Requires: C<xz> command in PATH

Result: C<filename.log.xz>

Uses multi-threaded compression (C<-T0>) and force overwrites temp file (C<-f>).

=head2 zstd

Requires: C<zstd> command in PATH

Result: C<filename.log.zst>

Uses multi-threaded compression (C<-T0>), quiet mode (C<-q>), and force
overwrites temp file (C<-f>).

=head1 SIDE EFFECTS

=over 4

=item * Moves files from SAFE_FAIL_DIR to SAFE_ARCHIVE_DIR

=item * Creates SAFE_ARCHIVE_DIR if it doesn't exist

=item * Creates SAFE_FAIL_DIR if it doesn't exist (for safety)

=item * Removes uncompressed files after successful compression

=item * Never deletes or overwrites existing archive files

=back

=head1 FILESYSTEM INTERACTIONS

=over 4

=item * B<Read:> Source directory (SAFE_FAIL_DIR)

=item * B<Write:> Destination directory (SAFE_ARCHIVE_DIR)

=item * B<Modify:> Removes source files after move (or compressed files after compression)

=item * B<Create:> Archive directory if missing, compressed files if requested

=back

=head1 PLATFORM NOTES

=head2 Unix-like (Linux, macOS)

- Uses File::Copy::move for atomic file moves
- External compression commands (xz, zstd) via system()
- Directory creation via File::Path::make_path

=head2 Windows

- File::Copy::move works on Windows NTFS
- External commands (xz.exe, zstd.exe) must be in PATH
- Path handling via File::Spec for portability

=head1 EXAMPLES

=head2 Archive All Logs (No Compression)

  $ safe-archive.pl --all
  ARCHIVED: .agent/FAIL-LOGS/fail-001.log -> .agent/FAIL-ARCHIVE/fail-001.log
  ARCHIVED: .agent/FAIL-LOGS/fail-002.log -> .agent/FAIL-ARCHIVE/fail-002.log

=head2 Archive with gzip Compression

  $ export SAFE_ARCHIVE_COMPRESS=gzip
  $ safe-archive.pl --all
  ARCHIVED: .agent/FAIL-LOGS/fail-001.log -> .agent/FAIL-ARCHIVE/fail-001.log
  # Compresses to: .agent/FAIL-ARCHIVE/fail-001.log.gz

=head2 Archive Specific File

  $ safe-archive.pl .agent/FAIL-LOGS/specific-fail.log
  ARCHIVED: .agent/FAIL-LOGS/specific-fail.log -> .agent/FAIL-ARCHIVE/specific-fail.log

=head2 Collision Handling

  $ ls .agent/FAIL-ARCHIVE/
  my-log.log
  
  $ safe-archive.pl .agent/FAIL-LOGS/my-log.log
  INFO: destination exists, using suffix: .agent/FAIL-ARCHIVE/my-log.1.log
  ARCHIVED: .agent/FAIL-LOGS/my-log.log -> .agent/FAIL-ARCHIVE/my-log.1.log

=head2 Custom Directories

  $ export SAFE_FAIL_DIR=/var/log/failures
  $ export SAFE_ARCHIVE_DIR=/archive/failures
  $ safe-archive.pl --all

=head2 xz Compression

  $ export SAFE_ARCHIVE_COMPRESS=xz
  $ safe-archive.pl --all
  # Creates .log.xz files with multi-threaded compression

=head1 CONTRACT REFERENCES

This script implements the archival behavior specified in:

=over 4

=item * M0-P1-I3: No-clobber semantics for safe archival

=item * docs/conformance-contract.md: Artifact management

=back

=head1 DIAGNOSTICS

=over 4

=item B<File not found: <path>>

The specified file does not exist. Check the path and try again.

=item B<Failed to move <src> -> <dest>: <error>>

File move operation failed. Common causes: permission denied, disk full,
source and destination on different filesystems (move requires copy+delete).

=item B<gzip compression requires IO::Compress::Gzip: <error>>

The IO::Compress::Gzip module is not installed. Install it via CPAN or
use a different compression method.

=item B<xz command not found in PATH>

The xz compression tool is not installed or not in PATH.

=item B<zstd command not found in PATH>

The zstd compression tool is not installed or not in PATH.

=item B<Invalid SAFE_ARCHIVE_COMPRESS value: <value>>

The compression method is not recognized. Use: none, gzip, xz, or zstd.

=back

=head1 SEE ALSO

=over 4

=item * L<safe-run.pl> - Create failure logs

=item * L<safe-check.pl> - Verify contract compliance

=item * L<IO::Compress::Gzip> - Perl gzip compression module

=back

=head1 AUTHOR

RFC-Shared-Agent-Scaffolding Project

=head1 LICENSE

Unlicense - See LICENSE file in repository root

=cut

use strict;
use warnings;
use File::Path qw(make_path);
use File::Basename qw(basename);
use File::Copy qw(move copy);

sub die_msg { print STDERR "ERROR: $_[0]\n"; exit 1; }

sub usage {
  print STDERR <<"USAGE";
Usage:
  scripts/perl/safe-archive.pl [--all | <file> ...]

Options:
  --all            Archive all files in the FAIL-LOGS directory.
  -h, --help       Show help.

Environment:
  SAFE_FAIL_DIR           Source directory (default: .agent/FAIL-LOGS)
  SAFE_ARCHIVE_DIR        Destination directory (default: .agent/FAIL-ARCHIVE)
  SAFE_ARCHIVE_COMPRESS   Compression: none|gzip|xz|zstd (default: none)

Notes:
  - Never deletes. Moves files without overwriting existing destination.
  - Compression is applied after move. gzip uses IO::Compress::Gzip if available; xz/zstd require external commands.
USAGE
  exit 2;
}

sub have_cmd {
  my ($cmd) = @_;
  for my $p (split(/:/, $ENV{PATH} // "")) {
    return 1 if -x "$p/$cmd";
  }
  return 0;
}

sub compress_file {
  my ($method, $path) = @_;
  $method ||= "none";
  return if $method eq "none";

  if ($method eq "gzip") {
    eval { require IO::Compress::Gzip; IO::Compress::Gzip->import(qw(gzip $GzipError)); 1 } or die_msg("gzip compression requires IO::Compress::Gzip: $@");
    my $out = "$path.gz";
    IO::Compress::Gzip::gzip($path => $out) or die_msg("gzip failed: $IO::Compress::Gzip::GzipError");
    unlink $path or die_msg("Failed to remove uncompressed after gzip: $!");
    return;
  }

  if ($method eq "xz") {
    die_msg("xz command not found in PATH") if !have_cmd("xz");
    system("xz", "-T0", "-f", $path) == 0 or die_msg("xz failed");
    return;
  }

  if ($method eq "zstd") {
    die_msg("zstd command not found in PATH") if !have_cmd("zstd");
    system("zstd", "-q", "-T0", "-f", $path) == 0 or die_msg("zstd failed");
    return;
  }

  die_msg("Invalid SAFE_ARCHIVE_COMPRESS value: $method (expected none|gzip|xz|zstd)");
}

my @argv = @ARGV;
if (!@argv || $argv[0] eq '-h' || $argv[0] eq '--help') { usage() if !@argv; usage(); }

my $fail_dir = $ENV{SAFE_FAIL_DIR} // ".agent/FAIL-LOGS";
my $archive_dir = $ENV{SAFE_ARCHIVE_DIR} // ".agent/FAIL-ARCHIVE";
my $compress = $ENV{SAFE_ARCHIVE_COMPRESS} // "none";

make_path($fail_dir) if !-d $fail_dir;
make_path($archive_dir) if !-d $archive_dir;

sub archive_one {
  my ($src) = @_;
  die_msg("File not found: $src") if !-e $src;
  my $base = basename($src);
  my $dest = "$archive_dir/$base";

  # M0-P1-I3: Auto-suffix if destination exists
  if (-e $dest) {
    my $suffix = 1;
    my ($name, $ext) = $base =~ /^(.+?)(\.[^.]+)?$/;
    $ext //= '';
    while (-e "$archive_dir/$name.$suffix$ext") {
      $suffix++;
    }
    $dest = "$archive_dir/$name.$suffix$ext";
    print STDERR "INFO: destination exists, using suffix: $dest\n";
  }
  
  move($src, $dest) or die_msg("Failed to move $src -> $dest: $!");
  print STDERR "ARCHIVED: $src -> $dest\n";
  compress_file($compress, $dest);
}

if (@argv && $argv[0] eq '--all') {
  shift @argv;
  opendir(my $dh, $fail_dir) or die_msg("Cannot open $fail_dir: $!");
  my @files = sort grep { -f "$fail_dir/$_" } readdir($dh);
  closedir($dh);
  if (!@files) {
    print STDERR "No files to archive in $fail_dir\n";
    exit 0;
  }
  for my $f (@files) { archive_one("$fail_dir/$f"); }
  exit 0;
}

usage() if !@argv;
for my $f (@argv) { archive_one($f); }
exit 0;
